syft/pkg/cataloger/golang/parse_go_bin.go

```diff
@@ -39,57 +67,121 @@ func newGoBinaryPackage(name, version, h1Digest, goVersion, architecture string,
+func getGOARCH(settings []debug.BuildSetting) string {
+	for _, s := range settings {
+		if s.Key == GOARCH {
+			return s.Value
+		}
 	}
-	return pkgs, err
+
+	return ""
 }
 
-func buildGoPkgInfo(location source.Location, mod, goVersion, arch string) []pkg.Package {
-	pkgsSlice := make([]pkg.Package, 0)
-	scanner := bufio.NewScanner(strings.NewReader(mod))
+func getGOARCHFromBin(r io.ReaderAt) (string, error) {
+	// Read the first bytes of the file to identify the format, then delegate to
+	// a format-specific function to load segment and section headers.
+	ident := make([]byte, 16)
+	if n, err := r.ReadAt(ident, 0); n < len(ident) || err != nil {
+		return "", errUnrecognizedFormat
+	}
 
-	// filter mod dependencies: [dep, name, version, sha]
-	for scanner.Scan() {
-		fields := strings.Fields(scanner.Text())
+	var arch string
+	switch {
+	case bytes.HasPrefix(ident, []byte("\x7FELF")):
+		f, err := elf.NewFile(r)
+		if err != nil {
+			return "", errUnrecognizedFormat
+		}
+		arch = f.Machine.String()
+	case bytes.HasPrefix(ident, []byte("MZ")):
+		f, err := pe.NewFile(r)
+		if err != nil {
+			return "", errUnrecognizedFormat
+		}
+		arch = fmt.Sprintf("%d", f.Machine)
+	case bytes.HasPrefix(ident, []byte("\xFE\xED\xFA")) || bytes.HasPrefix(ident[1:], []byte("\xFA\xED\xFE")):
+		f, err := macho.NewFile(r)
+		if err != nil {
+			return "", errUnrecognizedFormat
+		}
+		arch = f.Cpu.String()
+	case bytes.HasPrefix(ident, []byte{0x01, 0xDF}) || bytes.HasPrefix(ident, []byte{0x01, 0xF7}):
+		f, err := xcoff.NewFile(r)
+		if err != nil {
+			return "", errUnrecognizedFormat
+		}
+		arch = fmt.Sprintf("%d", f.FileHeader.TargetMachine)
+	default:
+		return "", errUnrecognizedFormat
+	}
+
+	arch = strings.Replace(arch, "EM_", "", 1)
+	arch = strings.Replace(arch, "Cpu", "", 1)
+	arch = strings.ToLower(arch)
+
+	return arch, nil
+}
+
+func getBuildSettings(settings []debug.BuildSetting) map[string]string {
+	m := make(map[string]string)
+	for _, s := range settings {
+		m[s.Key] = s.Value
+	}
+	return m
+}
+
+func buildGoPkgInfo(location source.Location, mod *debug.BuildInfo, arch string) []pkg.Package {
+	var pkgs []pkg.Package
+	if mod == nil {
+		return pkgs
+	}
 
-		// must have dep, name, version
-		if len(fields) < 3 {
+	for _, dep := range mod.Deps {
+		if dep == nil {
 			continue
 		}
 
-		name := fields[1]
-		version := fields[2]
-		h1Digest := ""
-		// if dep is *not* vendored, it'll also have h1digest
-		if len(fields) >= 4 {
-			h1Digest = fields[3]
-		}
+		pkgs = append(pkgs, newGoBinaryPackage(dep, mod.GoVersion, arch, location, nil))
+	}
 
-		if fields[0] == packageIdentifier {
-			pkgsSlice = append(pkgsSlice, newGoBinaryPackage(name, version, h1Digest, goVersion, arch, location))
-		}
-		if fields[0] == replaceIdentifier {
-			// replace the previous entry in the package slice
-			pkgsSlice[len(pkgsSlice)-1] = newGoBinaryPackage(name, version, h1Digest, goVersion, arch, location)
-		}
+	var empty debug.Module
```

Comment:
nit: this case isn't obvious when it would happen, a comment would be nice here to convey context to other developers